### PR TITLE
feat: publish

### PR DESCRIPTION
**Pull request by Jakub-Vacek**
## Description

This PR adds publish command.
Missing:
- [x] Login command
- [x] Check command scoped for one capability using .cache
- [x] Compile scoped for one capability using .cache
- [x] Lint scoped for one capability using .cache
- [x] Update tests when we have login merged

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have updated the documentation accordingly. For updating Oclif commands documentation use [oclif-dev](https://github.com/oclif/dev-cli#oclif-dev-readme).
- [x] I have read the **CONTRIBUTION_GUIDE** document.
- [x] I haven't repeated the code. (DRY)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/logic/publish.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+      );
+
+      if (!options?.dryRun) {
+        await client.createMap(file);
+      }
+    } else {
+      throw userError('Unknown map file structure', 1);
+    }
+  } else {
+    throw userError('Unknown file suffix', 1);
```
Comment (TheEdward162) [inline src/logic/publish.ts:80]: Extension is a better name

File: src/common/http.ts [hunk 2]
```diff
@@ -31,80 +34,132 @@ export enum ContentType {
...
+    }
+  );
+
+  await unwrapSuperfaceResponse(response);
 
-  return (response.body as { data: ProviderJson[] }).data;
+  return ((await response.json()) as { data: ProviderJson[] }).data;
 }
 
+//Unable to reuse service client getProfile due to profile ID resolution - version is always defined in service client
```
[the remark this answers is not in the document]
Comment (Jakub-Vacek) [reply src/common/http.ts:89]: Yep it should, but I don't know how exactly is Service Client used in Air etc. so I decided not to open PR and use what I can. Also there could be way to pass additional headers to ServiceClient methods (user-agent).

File: src/common/http.ts [hunk 3]
```diff
@@ -31,80 +34,132 @@ export enum ContentType {
...
-  return response.body as ProfileDocumentNode;
+  return (await response.json()) as ProfileDocumentNode;
 }
 
 export async function fetchProviderInfo(
   providerName: string
 ): Promise<ProviderJson> {
-  const query = new URL(providerName, `${getStoreUrl()}providers/`).href;
-  const response = await fetch(query, ContentType.JSON);
+  //TODO: user agent?
```
[the remark this answers is not in the document]
Comment (Jakub-Vacek) [reply src/common/http.ts:150]: I add it to my list of Service Client changes

**Review by TheEdward162 — COMMENTED** [review]
Currently ok with the draft

**Comment by Jakub-Vacek** [comment]
> Currently ok with the draft

Great, I need to finish Login(currently waiting on service client changes) and we should decide if we want to merge Check with Lint. If we decide not to merge it I will need lint to have similar user interface (scoped on one capability) and functions (use lint with any combination of local/remote map and profile, use .cache)

**Review by TheEdward162 — COMMENTED** [review]
I nitpicked on some english. Otherwise still ok 👍

File: README.md [hunk 4]
```diff
@@ -301,6 +302,41 @@ EXAMPLES
...
+## `superface publish PATH`
+
+Uploads map/profile/provider to Store. Published file must be locally linked in super.json. This command runs Check and Lint internaly to ensure quality
+
+```
+USAGE
+  $ superface publish PATH
+
+ARGUMENTS
+  PATH  (map|profile|provider) Document type of publeshed file
```
Comment (lukas-valenta) [inline README.md:315]: PATH?

File: src/common/http.ts [hunk 5]
```diff
@@ -170,9 +198,17 @@ export async function fetchMapAST(
...
+    method: 'GET',
+    headers: {
+      ...commonHeaders(),
+      Accept: ContentType.MAP_AST,
+    },
+  });
+
+  await checkSuperfaceResponse(response);
+
+  return (await response.json()) as MapDocumentNode;
```
Comment (lukas-valenta) [inline src/common/http.ts:213]: we have a validation function for this now

File: src/common/http.ts [hunk 6]
```diff
@@ -83,79 +86,104 @@ export function getServicesUrl(): string {
...
+    method: 'GET',
+    headers: {
+      ...commonHeaders(),
+      Accept: ContentType.PROFILE_AST,
+    },
+  });
+
+  await checkSuperfaceResponse(response);
+
+  return (await response.json()) as ProfileDocumentNode;
```
Comment (lukas-valenta) [inline src/common/http.ts:161]: this too

File: src/logic/lint.test.ts [hunk 7]
```diff
@@ -16,6 +15,7 @@ import {
   ValidationResult,
 } from '@superfaceai/parser';
 import { SyntaxErrorCategory } from '@superfaceai/parser/dist/language/error';
+import { MatchAttempts } from '@superfaceai/parser/dist/language/syntax/rule';
```
Comment (lukas-valenta) [inline src/logic/lint.test.ts:18]: why are these imports like this?

File: src/logic/publish.utils.ts [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+    throw userError(error, 1);
+  }
+
+  //Check map and profile
+  const result = checkMapAndProfile(profileAst, mapAst, options);
+
+  try {
+    parseProviderJson(providerJson);
+  } catch (error) {
+    //TODO: better way of formating?
```
Comment (lukas-valenta) [inline src/logic/publish.utils.ts:49]: this is super ugly, could you rather write a type guard?

File: src/logic/publish.utils.ts [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+): Promise<{ ast: ProfileDocumentNode; source?: string }> {
+  let ast: ProfileDocumentNode;
+
+  const source = await findLocalProfileSource(superJson, {
+    name: profile.name,
+    scope: profile.scope,
+    version,
+  });
+
+  const profileId = `${profile.scope ? `${profile.scope}/` : ''}${
```
Comment (lukas-valenta) [inline src/logic/publish.utils.ts:97]: this looks like something that should be a part of the ProfileId object

**Review by lukas-valenta — COMMENTED** [review]
i don't like the number of TODOs in this :confused:

File: src/logic/lint.test.ts (hunk 7 above)
Comment (Jakub-Vacek) [reply src/logic/lint.test.ts:18]: Because parser has weird exports 😩

Comment (Jakub-Vacek) [reply src/logic/lint.test.ts:18]: Should we fix this in Parser? This only effects the tests🤷

File: src/logic/publish.utils.ts (hunk 8 above)
Comment (Jakub-Vacek) [reply src/logic/publish.utils.ts]: Fixed it with checking if error is ZodError (I needed zod which is in sdk anyway so i added it to CLI).

File: src/logic/publish.utils.ts [hunk 10]
```diff
@@ -46,19 +47,17 @@ export function prePublishCheck(
...
-      for (const issue of (error as { issues: [] }).issues) {
-        if ('path' in issue && 'message' in issue) {
-          result.push({
-            kind: 'error',
-            message: `Provider check error: ${
-              (issue as { message: string }).message
-            } on path ${(issue as { path: string }).path}`,
-          });
-        }
+    if (error instanceof ZodError)
```
Comment (lukas-valenta) [inline src/logic/publish.utils.ts:50]: i meant just some simple typeguard, don't add the zod dependency which is going to be removed from sdk soon anyway

**Review by lukas-valenta — APPROVED** [review]

**Pull request merged**